src/Polly.Core/Retry/RetryResilienceStrategy.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using Polly.Strategy;
+using Polly.Telemetry;
+
+namespace Polly.Retry;
+
+internal class RetryResilienceStrategy : ResilienceStrategy
+{
+    private readonly TimeProvider _timeProvider;
+    private readonly ResilienceTelemetry _telemetry;
+    private readonly OutcomeEvent<OnRetryArguments, OnRetryEvent>.Handler? _onRetry;
+    private readonly OutcomeGenerator<TimeSpan, RetryDelayArguments, RetryDelayGenerator>.Handler? _delayGenerator;
+    private readonly OutcomePredicate<ShouldRetryArguments, ShouldRetryPredicate>.Handler? _shouldRetry;
+
+    public RetryResilienceStrategy(RetryStrategyOptions options, TimeProvider timeProvider, ResilienceTelemetry telemetry)
+    {
+        _timeProvider = timeProvider;
+        _telemetry = telemetry;
+        _onRetry = options.OnRetry.CreateHandler();
+        _delayGenerator = options.RetryDelayGenerator.CreateHandler();
+        _shouldRetry = options.ShouldRetry.CreateHandler();
+
+        BackoffType = options.BackoffType;
+        BaseDelay = options.BaseDelay;
+        RetryCount = options.RetryCount;
+    }
+
+    public TimeSpan BaseDelay { get; }
+
+    public RetryBackoffType BackoffType { get; }
+
+    public int RetryCount { get; }
+
+    protected internal override async ValueTask<TResult> ExecuteCoreAsync<TResult, TState>(Func<ResilienceContext, TState, ValueTask<TResult>> callback, ResilienceContext context, TState state)
+    {
+        if (_shouldRetry == null)
+        {
+            return await callback(context, state).ConfigureAwait(context.ContinueOnCapturedContext);
+        }
+
+        int attempt = 0;
+
+        while (true)
```

Comment:
Should this observe the `CancellationToken` in the context? `while (true)`'s always make me nervous 😄

Reply:
True, checking for cancellation. :)